### PR TITLE
Updates top-level doc-string of functional core Scopes

### PR DESCRIPTION
**Pull request by marcvanzee**
Thanks to @avital for some valuable input!

### CODE REVIEW

File: flax/core/scope.py [hunk 1]
```diff
@@ -14,13 +14,45 @@
...
+called "state" or "buffer"). We store such variables in collections as well.
+For example, a model using batch normalization may use the following collection:
+`{ "BatchNorm1": { "moving_mean": ..., "moving_average": ...} }`.
+
+The functional core Scopes defined below allow a programmer to have easy access
+to both trainable and non-trainable parameters within a specific (sub-)module.
+Scopes maintain "parallel pointers" to subtrees of different collections. For 
+instance within an instance of the BatchNorm layer, the programmer has easy 
+access to both the trainable parameters `scale` and `mean`, as well as the 
+non-trainable parameters `moving_mean` and `moving_average`.
```
Comment (jheek) [inline flax/core/scope.py:40]: I am missing the explanation of how users actually access params using scope. Shouldn't you have things like `scope.variable` and `scope.param` here?

**Review by jheek — CHANGES_REQUESTED** [review]
I have a few comments. I think we need to iterate on this some more to really make scopes easy to understand from the docstring

File: flax/core/scope.py [hunk 2]
```diff
@@ -14,13 +14,45 @@
...
+
+In addition, many types of models include other, non-parameter variables (also
+called "state" or "buffer"). We store such variables in collections as well.
+For example, a model using batch normalization may use the following collection:
+`{ "BatchNorm1": { "moving_mean": ..., "moving_average": ...} }`.
+
+The functional core Scopes defined below allow a programmer to have easy access
+to both trainable and non-trainable parameters within a specific (sub-)module.
+Scopes maintain "parallel pointers" to subtrees of different collections. For 
+instance within an instance of the BatchNorm layer, the programmer has easy
```
Comment (andsteing) [inline flax/core/scope.py:38]: Nit: "instance within an instance" - maybe "example within an instance" ?

**Comment by andsteing** [comment]
Linking this to #593

File: flax/core/scope.py [hunk 3]
```diff
@@ -14,13 +14,52 @@
...
+  def conv(scope, x, ...)::
+    weight = scope.params('weight', ...)
+    ...
+
+  def module_fn(scope, x, ...):
+    conv = scope.child(conv, 'conv')(x, ...)
+    ...
+
+The function `init` is then used to initialize a Scope, which transforms it into
+a function that expects RNG keys for all collections::
```
Comment (jheek) [inline flax/core/scope.py:56]: rng sequenced and collections aren't linked (except for paramaters).  But there is no 'batch_stats' rng

File: flax/core/scope.py [hunk 4]
```diff
@@ -14,13 +14,52 @@
...
+  vars = init(module_fn)({
+      'params': PRNGKey(0), 'batch_stats': PRNGKey(0)
+  }, input_shapes)
+
+Thus, the user provides rng seeds for all collections to the top-level scope,
```
Comment (jheek) [inline flax/core/scope.py:61]: to explain RNGs better we might need to show `scope.make_rng('some_name')`

File: flax/core/scope.py [hunk 5]
```diff
@@ -14,13 +14,52 @@
...
+    "batch_stats": {
+      "BatchNorm1": { "moving_mean": ..., "moving_average": ...}
+    }
+  }
+
+This variable dict has two so-called "collections": `params` (containing the 
+trainable parameters) and `batch_stats` (containing the --non-trainable-- batch
+statistics).
+
+The functional core Scope defined below allows easy access to all variables in a
```
Comment (jheek) [inline flax/core/scope.py:36]: nit: perhaps keep class names quoted for highlighting so Scope -> `Scope`

File: flax/core/scope.py [hunk 6]
```diff
@@ -14,13 +14,52 @@
...
+layer. For instance, from inside the BathcNorm1 layer, the variable "mean" is
+accessed as follows::
+  mean = scope.variable('batch_stats', 'moving_mean', jnp.zeros, (,))
+
+This will initialize it using `jnp.zeros` with shape `(,)`. The function
```
Comment (jheek) [inline flax/core/scope.py:41]: this shape here I find a bit confusing. `(,)` is the same as `()` which is a scalar. But batch stats are not scalars

File: flax/core/scope.py [hunk 7]
```diff
@@ -14,13 +14,52 @@
...
+parameters (e.g., batch statistics). For instance, the top-level variable
+dictionary of a Flax model using batch normalization may look as follows::
+  {
+    "params": {
+      "Conv1": { "weight": ..., "bias": ... },
+      "BatchNorm1": { "scale": ..., "mean": ... },
+      "Conv2": {...}
+    },
+    "batch_stats": {
+      "BatchNorm1": { "moving_mean": ..., "moving_average": ...}
```
Comment (jheek) [inline flax/core/scope.py:28]: nit: moving_mean and moving_average are the same thing. Normally we have ema_mean and ema_variance

**Review by jheek — CHANGES_REQUESTED** [review]

**Review by jheek — APPROVED** [review]

**Pull request merged**